src/ui/src/builder/composables/useDragToSelect.ts

```diff
@@ -0,0 +1,341 @@
+import { ref, computed, type Ref, onUnmounted } from "vue";
+import type { BuilderManagerMode } from "../builderManager";
+import { generateBuilderManager } from "../builderManager";
+
+type BuilderManager = ReturnType<typeof generateBuilderManager>;
+
+interface SelectionRectangle {
+	isSelecting: boolean;
+	startX: number;
+	startY: number;
+	left: number;
+	top: number;
+	width: number;
+	height: number;
+}
+
+const MIN_SELECTION_SIZE_PX = 5;
+const DRAG_SELECTION_CLICK_DELAY_MS = 100;
+
+interface UseDragToSelectOptions {
+	wrapperRef: Ref<HTMLElement | null>;
+	builderMode: Ref<BuilderManagerMode>;
+	builderManager: BuilderManager;
+	isAnnotating: Ref<boolean>;
+}
+
+export function useDragToSelect(options: UseDragToSelectOptions) {
+	const { wrapperRef, builderMode, builderManager, isAnnotating } = options;
+
+	const selectionRect = ref<SelectionRectangle>({
+		isSelecting: false,
+		startX: 0,
+		startY: 0,
+		left: 0,
+		top: 0,
+		width: 0,
+		height: 0,
+	});
+
+	const dragAbortController = new AbortController();
+
+	const isCursorSelecting = computed(() => selectionRect.value.isSelecting);
+
+	const isHoveringSelectableArea = ref(false);
+	const justCompletedDragSelection = ref(false);
+
+	function shouldAllowSelectionInBlueprints(target: HTMLElement): boolean {
+		const targetEl = target.closest<HTMLElement>("[data-writer-id]");
+		if (!targetEl) return true;
+
+		const blueprintsBlueprint = target.closest<HTMLElement>(
+			".BlueprintsBlueprint",
+		);
+		if (!blueprintsBlueprint) return false;
+
+		return targetEl === blueprintsBlueprint;
+	}
+
+	function isClickOnBlueprintsCanvas(target: HTMLElement): boolean {
+		const blueprintsBlueprint = target.closest<HTMLElement>(
+			".BlueprintsBlueprint",
+		);
+		if (!blueprintsBlueprint) return false;
+
+		const nodeContainer =
+			blueprintsBlueprint.querySelector<HTMLElement>(".nodeContainer");
+		return nodeContainer ? nodeContainer.contains(target) : true;
+	}
+
+	function cleanupDragSelection() {
+		if (!selectionRect.value.isSelecting) return;
+
+		selectionRect.value.isSelecting = false;
```

Comment:
If you use `shallowRef`, you'll need a full reassignment


```suggestion
		selectionRect.value = {
		  isSelecting: false,
	  	startX: 0,
	  	startY: 0,
	  	left: 0,
	  	top: 0,
	  	width: 0,
	  	height: 0,
	 };
```